frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py

```diff
@@ -0,0 +1,472 @@
+    ods_emitter = OdsEmitter(ods_file)
+    ods_emitter.print(ODS_BANNER)
+    impl_emitter = OdsImplEmitter(impl_file)
+    impl_emitter.print(ODS_IMPL_BANNER)
+    generator = OpGenerator(reg_ops, ods_emitter, impl_emitter)
+    generate_ops(generator)
+
+
+def generate_ops(g: "OpGenerator"):
+  # "Binary"-ops.
+  # There are some variation in these, so we spell them out in case if they
+  # need individual customization.
+  g.print_banner("Binary arithmetic ops")
+  g.std_binary_op("aten::add", "AddOp", "add")
+  g.std_binary_op("aten::atan2", "Atan2Op", "atan2")
+  g.std_binary_op("aten::div", "DivOp", "div")
+  g.std_binary_op("aten::floor_divide", "FloorDivideOp", "floor_divide")
+  g.std_binary_op("aten::mul", "MulOp", "mul")
+  g.std_binary_op("aten::remainder", "RemainderOp", "remainder")
+  g.std_binary_op("aten::true_divide", "TrueDivideOp", "true_divide")
+
+  # Unary-ops. These are all the same so just name munge them.
+  g.print_banner("Unary arithmetic ops")
+  for uname in [
+      "abs", "acos", "angle", "asin", "atan", "ceil", "conj", "cos", "cosh",
+      "digamma", "erf", "erfc", "erfinv", "exp", "expm1", "floor", "frac",
+      "lgamma", "log", "log10", "log1p", "log2", "neg", "relu", "reciprocal",
+      "round", "rsqrt", "sigmoid", "sign", "sin", "sinh", "sqrt", "tan", "tanh",
+      "trunc"
+  ]:
+    g.std_unary_op(f"aten::{uname}", f"{snakecase_to_camelcase(uname)}Op",
+                   uname)
+
+
+def dump_registered_ops(outfile, reg_ops_dict):
+  for k in sorted(reg_ops_dict.keys()):
+    attr_dict = reg_ops_dict[k]
+    outfile.write(f"OP '{k}':\n")
+    for attr_name, attr_value in attr_dict.items():
+      outfile.write(f"    {attr_name} = {attr_value!r}\n")
+    outfile.write("\n")
+
+
+class OpGenerator:
+
+  def __init__(self, reg_ops, ods_emitter, impl_emitter):
+    super().__init__()
+    self.reg_ops = reg_ops
+    self.ods_emitter = ods_emitter
+    self.impl_emitter = impl_emitter
+
+  def print_banner(self, text):
+    for em in (self.ods_emitter, self.impl_emitter):
+      em.print(
+          "// -----------------------------------------------------------------------------"
+      )
+      em.print(f"// {text}")
+      em.print(
+          "// -----------------------------------------------------------------------------"
+      )
+      em.print("")
+
+  def std_binary_op(self, reg_name, ods_name, op_name):
+    """"Binary"-ops. These ops typically have:
+      - '.Tensor' variant where the second arg is a Tensor
+      - '.Scalar' variant where the second arg is a Scalar
+      - An '.out' variant which contains a final "outref" argument
+    Actual suffixes vary and the rules are set up to match anything
+    that comes in in one of these forms.
+    In addition, most of these have in-place versions, possibly of the
+    '.Tensor' and '.Scalar' variants above.
+    Note that many of these have more than two arguments (i.e. alpha/beta
```

Comment:
unclear how having additional arguments isn't load-bearing? How can we ignore part of the signature?

Reply:
Let me word-smith it. We process the whole signature but only have special rules around the first two arguments, which are what varies. The rest are constant-ey things and carried verbatim. If that ever doesn't apply, we just fork the offending op and do it manually or as a one off.